File: mcp-server/src/test_notes_store.py
"""
Unit tests for notes_store: the path fence and the persistence round-trip.

Run from the mcp-server/ folder:
    python src/test_notes_store.py

The fence tests feed deliberate traversal payloads (``../..`` chains, absolute
paths) and assert each is refused with a ValueError that names the path -- not
silently resolved, not silently returning nothing. Each refusal is printed so
you can see exactly what the caller would get back.

Exits 0 when the fence holds and persistence works, 1 (with a message) on any
regression -- same gating role as test_connection.py.
"""

import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from notes_store import (  # noqa: E402  (path insert must come first)
    DATA_ROOT,
    NOTES_FILENAME,
    load_notes,
    resolve_within,
    save_notes,
)


def _run(name, fn):
    try:
        fn()
    except AssertionError as exc:
        print(f"FAIL: {name}: {exc}", file=sys.stderr)
        sys.exit(1)
    print(f"PASS: {name}")


def test_allows_the_notes_file():
    resolved = resolve_within(DATA_ROOT, NOTES_FILENAME)
    assert resolved == DATA_ROOT.resolve() / NOTES_FILENAME
    assert resolved.is_relative_to(DATA_ROOT.resolve())


def test_rejects_dotdot_traversal():
    payloads = [
        "../../etc/hosts",
        "../../../../etc/passwd",
        "notes/../../../secret.txt",
    ]
    for payload in payloads:
        try:
            resolve_within(DATA_ROOT, payload)
        except ValueError as exc:
            print(f"       refused {payload!r} -> {exc}")
            continue
        raise AssertionError(f"{payload!r} was NOT refused")


def test_rejects_absolute_path_outside_root():
    outside = "C:\\Windows\\win.ini" if sys.platform == "win32" else "/etc/hosts"
    try:
        resolve_within(DATA_ROOT, outside)
    except ValueError as exc:
        print(f"       refused {outside!r} -> {exc}")
        return
    raise AssertionError(f"{outside!r} was NOT refused")


def test_persistence_round_trip():
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        seeded = load_notes(root)
        assert (root / NOTES_FILENAME).exists(), "first load did not create the store file"
        assert isinstance(seeded, list) and seeded, "seeded store should be a non-empty list"

        seeded.append(
            {"id": "marker", "title": "persisted", "content": "x", "created_at": "now"}
        )
        save_notes(seeded, root)

        reloaded = load_notes(root)
        assert any(n["id"] == "marker" for n in reloaded), "saved note did not survive reload"


def test_corrupt_store_errors_loudly():
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        (root / NOTES_FILENAME).write_text("{ not valid json", encoding="utf-8")
        try:
            load_notes(root)
        except RuntimeError as exc:
            print(f"       corrupt store raised -> {exc}")
            return
        raise AssertionError("corrupt store was read as empty instead of raising")


if __name__ == "__main__":
    _run("fence allows the notes file", test_allows_the_notes_file)
    _run("fence rejects ../ traversal", test_rejects_dotdot_traversal)
    _run("fence rejects absolute path outside root", test_rejects_absolute_path_outside_root)
    _run("persistence round-trips", test_persistence_round_trip)
    _run("corrupt store errors loudly", test_corrupt_store_errors_loudly)
    print("\nALL PASS: fence holds, persistence works.")

File: mcp-server/nodejs/test-notesStore.js
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import {
  DATA_ROOT,
  NOTES_FILENAME,
  loadNotes,
  resolveWithin,
  saveNotes,
} from "./notesStore.js";

// Unit tests for notesStore: the path fence and the persistence round-trip.
// Mirrors ../src/test_notes_store.py. No test framework — assertions and an
// exit code, so it can gate a build. Each refusal is printed so you can see
// exactly what a caller would get back.
//
// Run from the nodejs/ folder:  node test-notesStore.js   (or: npm run test:store)

function run(name, fn) {
  try {
    fn();
  } catch (err) {
    console.error(`FAIL: ${name}: ${err.message}`);
    process.exit(1);
  }
  console.log(`PASS: ${name}`);
}

run("fence allows the notes file", () => {
  const resolved = resolveWithin(DATA_ROOT, NOTES_FILENAME);
  assert.equal(resolved, path.join(fs.realpathSync(path.dirname(DATA_ROOT)), path.basename(DATA_ROOT), NOTES_FILENAME));
});

run("fence rejects ../ traversal", () => {
  const payloads = [
    "../../etc/hosts",
    "../../../../etc/passwd",
    "notes/../../../secret.txt",
  ];
  for (const payload of payloads) {
    let refused = false;
    try {
      resolveWithin(DATA_ROOT, payload);
    } catch (err) {
      refused = true;
      console.log(`       refused ${JSON.stringify(payload)} -> ${err.message}`);
    }
    assert.ok(refused, `${JSON.stringify(payload)} was NOT refused`);
  }
});

run("fence rejects absolute path outside root", () => {
  const outside = process.platform === "win32" ? "C:\\Windows\\win.ini" : "/etc/hosts";
  let refused = false;
  try {
    resolveWithin(DATA_ROOT, outside);
  } catch (err) {
    refused = true;
    console.log(`       refused ${JSON.stringify(outside)} -> ${err.message}`);
  }
  assert.ok(refused, `${JSON.stringify(outside)} was NOT refused`);
});

run("persistence round-trips", () => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), "mcp-notes-"));
  try {
    const seeded = loadNotes(root);
    assert.ok(fs.existsSync(path.join(root, NOTES_FILENAME)), "first load did not create the store file");
    assert.ok(Array.isArray(seeded) && seeded.length > 0, "seeded store should be a non-empty array");

    seeded.push({ id: "marker", title: "persisted", content: "x", createdAt: "now" });
    saveNotes(seeded, root);

    const reloaded = loadNotes(root);
    assert.ok(reloaded.some((n) => n.id === "marker"), "saved note did not survive reload");
  } finally {
    fs.rmSync(root, { recursive: true, force: true });
  }
});

run("corrupt store errors loudly", () => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), "mcp-notes-"));
  try {
    fs.writeFileSync(path.join(root, NOTES_FILENAME), "{ not valid json", "utf8");
    let raised = false;
    try {
      loadNotes(root);
    } catch (err) {
      raised = true;
      console.log(`       corrupt store raised -> ${err.message}`);
    }
    assert.ok(raised, "corrupt store was read as empty instead of throwing");
  } finally {
    fs.rmSync(root, { recursive: true, force: true });
  }
});

console.log("\nALL PASS: fence holds, persistence works.");
